### PR TITLE
Migrate TextTheme to 2018 APIs

### PR DESCRIPTION
**Pull request by HansMuller**
This PR is a breaking change. It's described in additional detail here:  flutter.dev/go/update-text-theme-api. It's based on an earlier version - https://github.com/flutter/flutter/pull/45745.

The TextTheme's styles now have names that match the [2018 Material Design spec](https://material.io/design/typography/the-type-system.html). Getters and parameters with the original names are still supported, but they are deprecated.

```
OLD NAME      NEW NAME
----------------------
display4      headline1
display3      headline2
display2      headline3
display1      headline4
headline      headline5
title         headline6
subhead       subtitle1
body2         bodyText1
body1         bodyText2
caption       caption
button        button
subtitle      subtitle2
overline      overline
```

The text styles called "bodyText1" and "bodyText2" are called body1 and body2 in the spec, H1-H6 are called headline1-headline6 in the TextTheme API:

image

Note that this PR only changes the names of the TextTheme TextStyles, it does not change the default text style configurations. By default, the text style configurations (the font size, font weight, letter spacing, etc), still match their original 2014 values. Apps can choose to use the 2018 text style configurations instead, by specifying Typography.material2018() as the value for their Theme's [typography](https://api.flutter.dev/flutter/material/ThemeData/typography.html).

### CODE REVIEW

File: examples/flutter_gallery/lib/demo/typography_demo.dart [hunk 1]
```diff
@@ -49,15 +49,15 @@ class TypographyDemo extends StatelessWidget {
...
-        TextStyleItem(name: 'Display 4', style: textTheme.display4, text: 'Light 112sp'),
-      TextStyleItem(name: 'Display 3', style: textTheme.display3, text: 'Regular 56sp'),
-      TextStyleItem(name: 'Display 2', style: textTheme.display2, text: 'Regular 45sp'),
-      TextStyleItem(name: 'Display 1', style: textTheme.display1, text: 'Regular 34sp'),
-      TextStyleItem(name: 'Headline', style: textTheme.headline, text: 'Regular 24sp'),
-      TextStyleItem(name: 'Title', style: textTheme.title, text: 'Medium 20sp'),
-      TextStyleItem(name: 'Subheading', style: textTheme.subhead, text: 'Regular 16sp'),
-      TextStyleItem(name: 'Body 2', style: textTheme.body2, text: 'Medium 14sp'),
-      TextStyleItem(name: 'Body 1', style: textTheme.body1, text: 'Regular 14sp'),
+        TextStyleItem(name: 'Headline 1', style: textTheme.headline1, text: 'Light 112sp'),
```
Comment (rami-a) [inline examples/flutter_gallery/lib/demo/typography_demo.dart]: Should the typography demo use the 2018 values for font weight and size instead now?

File: packages/flutter/lib/src/material/typography.dart [hunk 2]
```diff
@@ -256,7 +320,7 @@ class Typography extends Diagnosticable {
   @override
   void debugFillProperties(DiagnosticPropertiesBuilder properties) {
     super.debugFillProperties(properties);
-    final Typography defaultTypography = Typography();
+    final Typography defaultTypography = Typography.material2018();
```
Comment (rami-a) [inline packages/flutter/lib/src/material/typography.dart]: Does making the default 2018 here change anything?

**Review by rami-a — APPROVED** [review]
```
 __    ___  ____  __  __ 
(  )  / __)(_  _)(  \/  )
 )(__( (_-.  )(   )    ( 
(____)\___/ (__) (_/\/\_)
```

This is awesome!

File: examples/flutter_gallery/lib/demo/typography_demo.dart (hunk 1 above)
Comment (HansMuller) [reply examples/flutter_gallery/lib/demo/typography_demo.dart]: That's a good idea although an update to the typography demo would be probably be best as separate PR

File: packages/flutter/lib/src/material/typography.dart (hunk 2 above)
Comment (HansMuller) [reply packages/flutter/lib/src/material/typography.dart]: Good catch! Currently the default value for is the 2014 text theme.

File: examples/flutter_gallery/lib/demo/typography_demo.dart [hunk 3]
```diff
@@ -49,15 +49,15 @@ class TypographyDemo extends StatelessWidget {
...
+      TextStyleItem(name: 'Headline 2', style: textTheme.headline2, text: 'Regular 56sp'),
+      TextStyleItem(name: 'Headline 3', style: textTheme.headline3, text: 'Regular 45sp'),
+      TextStyleItem(name: 'Headline 4', style: textTheme.headline4, text: 'Regular 34sp'),
+      TextStyleItem(name: 'Headline 5', style: textTheme.headline5, text: 'Regular 24sp'),
+      TextStyleItem(name: 'Headline 6', style: textTheme.headline6, text: 'Medium 20sp'),
+      TextStyleItem(name: 'Subtitle 1', style: textTheme.subtitle1, text: 'Regular 16sp'),
```
Comment (Piinks) [inline examples/flutter_gallery/lib/demo/typography_demo.dart]: Is `Subtitle 2` missing here? I don't know if it is necessary, I noticed `overline` is also omitted.

Comment (HansMuller) [reply examples/flutter_gallery/lib/demo/typography_demo.dart]: Yes, both of those styles are missing. This is just a simple translation of the existing demo - only the names have changed - I'm going to update to the demo to show the complete set of 2018 text styles in a separate PR.

https://github.com/flutter/flutter/pull/48547#discussion_r366971531

File: packages/flutter/test/material/theme_test.dart [hunk 4]
```diff
@@ -376,15 +376,15 @@ void main() {
...
-        textTheme.body1,
+        textTheme.headline1,
+        textTheme.headline2,
+        textTheme.headline3,
+        textTheme.headline4,
+        textTheme.headline5,
+        textTheme.headline6,
+        textTheme.subtitle1,
+        textTheme.bodyText1,
+        textTheme.bodyText2,
```
Comment (Piinks) [inline packages/flutter/test/material/theme_test.dart]: Is subtitle(14)/subtitle2(18) missing here?

Comment (HansMuller) [reply packages/flutter/test/material/theme_test.dart]: This test just checks the original 2014 text styles (still the default). It's now a regression test. Will add a comment to that effect.

File: packages/flutter/lib/src/material/typography.dart [hunk 5]
```diff
@@ -364,37 +428,37 @@ class Typography extends Diagnosticable {
...
+    headline6 : TextStyle(debugLabel: 'dense title 2014',    inherit: false, fontSize:  21.0, fontWeight: FontWeight.w500, textBaseline: TextBaseline.ideographic),
+    subtitle1 : TextStyle(debugLabel: 'dense subhead 2014',  inherit: false, fontSize:  17.0, fontWeight: FontWeight.w400, textBaseline: TextBaseline.ideographic),
+    bodyText1 : TextStyle(debugLabel: 'dense body2 2014',    inherit: false, fontSize:  15.0, fontWeight: FontWeight.w500, textBaseline: TextBaseline.ideographic),
+    bodyText2 : TextStyle(debugLabel: 'dense body1 2014',    inherit: false, fontSize:  15.0, fontWeight: FontWeight.w400, textBaseline: TextBaseline.ideographic),
+    caption   : TextStyle(debugLabel: 'dense caption 2014',  inherit: false, fontSize:  13.0, fontWeight: FontWeight.w400, textBaseline: TextBaseline.ideographic),
+    button    : TextStyle(debugLabel: 'dense button 2014',   inherit: false, fontSize:  15.0, fontWeight: FontWeight.w500, textBaseline: TextBaseline.ideographic),
+    subtitle2 : TextStyle(debugLabel: 'dense subtitle 2014', inherit: false, fontSize:  15.0, fontWeight: FontWeight.w500, textBaseline: TextBaseline.ideographic),
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/typography.dart]: I reordered the typography styles; now they match.

File: packages/flutter/lib/src/material/typography.dart [hunk 6]
```diff
@@ -268,94 +332,94 @@ class Typography extends Diagnosticable {
...
+    headline4 : TextStyle(debugLabel: 'blackMountainView headline4', fontFamily: 'Roboto', inherit: true, color: Colors.black54, decoration: TextDecoration.none),
+    headline5 : TextStyle(debugLabel: 'blackMountainView headline5', fontFamily: 'Roboto', inherit: true, color: Colors.black87, decoration: TextDecoration.none),
+    headline6 : TextStyle(debugLabel: 'blackMountainView headline6', fontFamily: 'Roboto', inherit: true, color: Colors.black87, decoration: TextDecoration.none),
+    subtitle1 : TextStyle(debugLabel: 'blackMountainView subtitle1', fontFamily: 'Roboto', inherit: true, color: Colors.black87, decoration: TextDecoration.none),
+    bodyText1 : TextStyle(debugLabel: 'blackMountainView bodyText1', fontFamily: 'Roboto', inherit: true, color: Colors.black87, decoration: TextDecoration.none),
+    bodyText2 : TextStyle(debugLabel: 'blackMountainView bodyText2', fontFamily: 'Roboto', inherit: true, color: Colors.black87, decoration: TextDecoration.none),
+    caption   : TextStyle(debugLabel: 'blackMountainView caption',   fontFamily: 'Roboto', inherit: true, color: Colors.black54, decoration: TextDecoration.none),
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/typography.dart]: Not intentional, thanks for pointing that out.

File: packages/flutter/lib/src/material/text_theme.dart [hunk 7]
```diff
@@ -482,57 +714,57 @@ class TextTheme extends Diagnosticable {
...
+    properties.add(DiagnosticsProperty<TextStyle>('headline1', headline1, defaultValue: defaultTheme.headline1));
+    properties.add(DiagnosticsProperty<TextStyle>('headline2', headline2, defaultValue: defaultTheme.headline2));
+    properties.add(DiagnosticsProperty<TextStyle>('headline3', headline3, defaultValue: defaultTheme.headline3));
+    properties.add(DiagnosticsProperty<TextStyle>('headline4', headline4, defaultValue: defaultTheme.headline4));
+    properties.add(DiagnosticsProperty<TextStyle>('headline5', headline5, defaultValue: defaultTheme.headline5));
+    properties.add(DiagnosticsProperty<TextStyle>('headline6', headline6, defaultValue: defaultTheme.headline6));
+    properties.add(DiagnosticsProperty<TextStyle>('subtitle1', subtitle1, defaultValue: defaultTheme.subtitle1));
+    properties.add(DiagnosticsProperty<TextStyle>('subtitle2', subtitle2, defaultValue: defaultTheme.subtitle2));
     properties.add(DiagnosticsProperty<TextStyle>('body1', body1, defaultValue: defaultTheme.body1));
+    properties.add(DiagnosticsProperty<TextStyle>('body2', body2, defaultValue: defaultTheme.body2));
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/text_theme.dart]: Yes!

File: packages/flutter/lib/src/material/text_theme.dart [hunk 8]
```diff
@@ -482,57 +714,57 @@ class TextTheme extends Diagnosticable {
...
+      headline1,
+      headline2,
+      headline3,
+      headline4,
+      headline5,
+      headline6,
+      subtitle1,
+      subtitle2,
       body1,
+      body2,
```
[the remark this answers is not in the document]
Comment (HansMuller) [reply packages/flutter/lib/src/material/text_theme.dart]: Yes!

File: packages/flutter/lib/src/material/typography.dart (hunk 6 above)
[the remark this answers is not in the document]
Comment (Piinks) [reply packages/flutter/lib/src/material/typography.dart]: Ah, it looks like the whitespace is meant to make this a whole lot more readable. I guess it just depends where you are viewing it from. :)
(image not from this change)
Screen Shot 2020-01-24 at 11 04 54 AM

**Review by Piinks — APPROVED** [review]
Flutter_LGTM
Well done! This is a behemoth! :)

File: packages/flutter/lib/src/material/text_theme.dart [hunk 9]
```diff
@@ -210,33 +385,90 @@ class TextTheme extends Diagnosticable {
...
+      'This feature was deprecated after v1.13.8.'
+    )
+    TextStyle subtitle,
+    @Deprecated(
+      'This is the term used in the 2014 version of material design. The modern term is bodyText1. '
+      'This feature was deprecated after v1.13.8.'
+    )
     TextStyle body2,
+    @Deprecated(
+      'This is the term used in the 2014 version of material design. The modern term is bodyText2. '
```
Comment (Hixie) [inline packages/flutter/lib/src/material/text_theme.dart]: The "bodyText2" here should be "bodyText1" and vice versa.

**Pull request merged**